### ISSUE TITLE
support creation of multiple DAG instances with one configuration file (prerequisite for pipetest)

### ISSUE THREAD

**Issue by windiana42** [pydiverse/pydiverse.pipedag#17]
The power of pipedag is that it should make it easy to create multiple instances of a DAG working with small/medium/large data inputs and per-user as well as team shared materialization. However, we still need to define and implement a configuration format that allows specifying multiple DAG instances. I would like to keep the binding to pipetest loose by just defining tags that may or may not be picked up by pipetest to schedule DAG instance executions with a unit testing framework.
An tricky point is that sometimes one likes to develop on some stages of a pipeline which are technically cache-invalid in case of new source input. But for development purposes one might want to ignore this fresh input.

**Comment by windiana42** [comment]
@NMAC427 @verticube here is my first attempt to convert the existing pipedag configuration (https://github.com/pydiverse/pydiverse.pipedag/blob/main/tests/pipedag.toml, https://github.com/pydiverse/pydiverse.pipedag/blob/main/docs/reference/config.rst) to one that may be used for defining the execution of multiple instances:
```
name: pipedag_tests
database_connections:
  postgres:
    # Postgres: this can be used after running `docker-compose up`
    engine: "postgresql://sa:Pydiverse23@127.0.0.1:6543/{database}"

  mssql:
    # SQL Server: this can be used after running `docker-compose up`
    engine: "mssql+pyodbc://sa:PidyQuant27@127.0.0.1:1433/master?driver=ODBC+Driver+18+for+SQL+Server&encrypt=no"
    # schema_prefix: "master."  # SQL Server needs database.schema (uncomment only on of prefix and suffix)
    schema_prefix: "{database}_"  # SQL Server needs database.schema (uncomment only on of prefix and suffix)
    schema_suffix: ".dbo"   # Alternatively SQL Server databases can be used as schemas with .dbo default schema

instances:
  default:
    # listen-interface for pipedag context server which synchronizes some task state during DAG execution
    network_interface: "127.0.0.1"
    # classes to be materialized to table store even without pipedag Table wrapper (we have loose coupling between
    # pipedag and pydiverse.transform, so consider adding 'pydiverse.transform.Table' in your config)
    auto_table: ["pandas.DataFrame", ]
    # abort as fast a possible on task failure and print most readable stack trace
    fail_fast: true

    table_store:
      class: "pydiverse.pipedag.backend.table.SQLTableStore"

      # Postgres: this can be used after running `docker-compose up`
      database_connection: postgres
      database: pipedag_default

      ## SQL Server: this can be used after running `docker-compose up`
      #database: mssql

      # print select statements before being encapsualted in materialize expressions and tables before writing to
      # database
      print_materialize: true
      # print final sql statements
      print_sql: true

    blob_store:
      class: "pydiverse.pipedag.backend.blob.FileBlobStore"
      base_path: "/tmp/pipedag/blobs"

    lock_manager:
      class: "pydiverse.pipedag.backend.lock.ZooKeeperLockManager"
      hosts: "localhost:2181"

    engine:
      class: "pydiverse.pipedag.engine.SequentialEngine"
      ## Activate this class to work either with prefect 1.3.0 or prefect 2.0
      # class: "pydiverse.pipedag.engine.PrefectEngine"
    # Flow attributes are handed over to the flow implementation (pipedag does not care)
    flow_attributes:
      # by default we load source data and not a sampled version of a loaded database
      copy_filtered_input: false

  full:
    # Full dataset is using default database connection and schemas
    table_store:
      database: pipedag_full
    # Run this instance under @pytest.mark.slow5
    tag: pytest_slow5

  midi:
    # Full dataset is using default database connection and schemas
    table_store:
      database: pipedag_midi
    flow_attributes:
      # copy filtered input from full instance
      copy_filtered_input: true
      copy_source: pipedag_full
      sample_cnt: 2  # this is just dummy input where we sample 2 rows
    # Run this instance under @pytest.mark.slow4
    tag: pytest_slow4
    # Run only stage_2 under @pytest.mark.slow3
    stage_tags:
      pytest_slow3:
        - stage_2

  mini:
    # Full dataset is using default database connection and schemas
    table_store:
      database: pipedag_mini
    flow_attributes:
      copy_filtered_input: true
      copy_source: pipedag_full
      sample_cnt: 1  # this is just dummy input where we sample 1 row
    # Run this instance under @pytest.mark.slow2
    tag: pytest_slow2
    # Run only stage_2 under @pytest.mark.slow1
    stage_tags:
      pytest_slow1:
        - stage_2
```

**Comment by windiana42** [comment]
@NMAC427 an aspect I also like to change is that configuration should be available directly after parsing command line arguments and not just within flow.run(). It could look like this:
```python
@click ...
def main(pipedag_instance, per_user, ignore_fresh_input)
  cfg = pipedag.read_config(pipedag_instance, per_user, ignore_fresh_input)
  with Flow(cfg):
    if cfg.attributes["copy_filtered_input"]:
      a = copy_input()  # S1 for ILmig
    else:
      a = input()  # S0 for ILmig
    ...
```

**Comment by windiana42** [comment]
For testing as entrypoint to flow execution, I am unsure whether it is possible to create an annotation that can create different test functions (like @parameterize) but with allocation to different @pytest.mark.xxx markers.
```python
@pipetest.run(instances=["mini", "midi", "full"], per_user=False)
def test_flow(pipetest_runner, cfg)
  pipetest_runner.run(mypackage.get_flow(cfg))
```
Alternatively, one could also add a top level function call that declares multiple functions either in conftest.py or in test files. I would prefer the syntax above if feasible since it is less magic in my opinion.

**Comment by windiana42** [comment]
It would still be nice to support normal unit tests where each defines a different flow and thus the instance definitions don't make that much sense:
```python
@fixture(scope=module)
def instance_config():
  return pipedag.read_config()

def test_flow(instance_config)
  cfg = instance_config.get(instance="full", per_user=True, ignore_fresh_input)
  flow = mypackage.get_flow(cfg)
  result = flow.run()
  assert result.successful
```

**Comment by windiana42** [comment]
@NMAC427 we have a collision with the term `instance`:
```python
def load_instance(config_dict: dict):
    """Instantiates an instance of an object given
...
```
When I talk about instance in this issue, I mean a pipe-DAG instance.

**Comment by windiana42** [comment]
@NMAC427 this is a neat code construct in ConfigContext, but I like config interpretation errors to occur before first use:
```python
    @cached_property
    def store(self) -> PipeDAGStore:
        from pydiverse.pipedag.materialize.store import PipeDAGStore

        table_store = load_object(self.config_dict["table_store"])
        ...
```

**Comment by windiana42** [comment]
I would also like to rename the config parameter `engine` to `dag_engine`. The second `engine` parameter we have in database_connections might be renamed as `connection_string` or simply `url`.

**Comment by NMAC427** [comment]
> An aspect I also like to change is that configuration should be available directly after parsing command line arguments and not just within flow.run().


The reason the config is not just an object but a context manager is because the config should ideally be independent from the flow definition. This means that you should be able to run the same flow object with different config instances. Your proposal, however, tightly couples a configuration with a flow. This means once a flow has been instantiated with a specific config, you can't run it with a different config.

Instead you should be able to achieve the same result by accessing the ConfigContext inside a task and return a different output based on it:

```python

@materialize(cache=never)
def data_source():
    config = ConfigContext.get()
    if config.attributes["copy_filtered_input"]:
        return "copy-filtered"
    return "read-from-disk"


@materialize
def input(data_source):
    if data_source == "copy-filtered":
        ...
    elif data_source == "read-from-disk":
        ...


with Flow():
    data_src = data_source()
    a = input(data_src)
    b = clean(a)

# To choose a specific "instance" I propose this syntax
flow.run(with_instance="mini")
flow.run(with_instance="full")

# To run a flow with a completely different config file, you should also be
# able to open a ConfigContext yourself.
with ConfigContext.fromFile("/a/very/different/config/file.yaml"):
    flow.run(with_instance="full")
```

---

> When I talk about instance in this issue, I mean a pipe-DAG instance. Thus I would like to rename the method above `load_object`.

`init_object` might be the best name, because loading usually refers to some kind of data transfer.

---

> I would also like to rename the config parameter `engine` to `dag_engine`, `orchestration_engine`, or `orchestration`. 

I believe sticking with the word engine makes most sense. I see how it potentially could be confused with the database engine, however, by simply looking at an example config file it should be clear to anyone that it refers to the flow execution back end. Simply sticking with engine also makes sense, because the corresponding namespace is `pydiverse.pipedag.engine`.


> The second `engine` parameter we have in database_connections might be renamed as `connection_string` or simply `url`.

I like this idea, and I would go with `url`, as it corresponds to the argument name used by sqlalchemy in the `create_engine` function.


---


Finally, I think we need to define the term "instance" more thoroughly.
- Is it a different variation of the config file?
- Is it a flow object with different arguments?
- Is it a flow + a specific configuration?
- Is it a specific "run" of a flow?

Based on the example config file you provided it seems like an instance refers to a different variation of the config file (E.g. the instance "full" connects to a different database as "mini").

**Comment by windiana42** [comment]
This is my current best guess for a pipedag configuration file:
```yaml
name: pipedag_tests
strict_instance_lookup: true  # default value: true
database_connections:
  postgres:
    # Postgres: this can be used after running `docker-compose up`
    url: "postgresql://sa:Pydiverse23@127.0.0.1:6543/{instance_id}"

  mssql:
    # SQL Server: this can be used after running `docker-compose up`
    url: "mssql+pyodbc://sa:PidyQuant27@127.0.0.1:1433/master?driver=ODBC+Driver+18+for+SQL+Server&encrypt=no"
    # schema_prefix: "master."  # SQL Server needs database.schema (uncomment only on of prefix and suffix)
    schema_prefix: "{instance_id}_"  # SQL Server needs database.schema (uncomment only on of prefix and suffix)
    schema_suffix: ".dbo"   # Alternatively SQL Server databases can be used as schemas with .dbo default schema

blob_connections:
  file:
    base_path: "/tmp/pipedag/blobs"

technical_setups:
  default:
    # listen-interface for pipedag context server which synchronizes some task state during DAG execution
    network_interface: "127.0.0.1"
    # classes to be materialized to table store even without pipedag Table wrapper (we have loose coupling between
    # pipedag and pydiverse.transform, so consider adding 'pydiverse.transform.Table' in your config)
    auto_table: ["pandas.DataFrame", ]
    # abort as fast a possible on task failure and print most readable stack trace
    fail_fast: true

    instance_id: pipedag_default
    table_store:
      class: "pydiverse.pipedag.backend.table.SQLTableStore"

      # Postgres: this can be used after running `docker-compose up`
      database_connection: postgres
      create_database_if_not_exists: True

      ## SQL Server: this can be used after running `docker-compose up`
      #database_connection: mssql

      # print select statements before being encapsualted in materialize expressions and tables before writing to
      # database
      print_materialize: true
      # print final sql statements
      print_sql: true

    blob_store:
      class: "pydiverse.pipedag.backend.blob.FileBlobStore"
      blob_connections: file

    lock_manager:
      class: "pydiverse.pipedag.backend.lock.ZooKeeperLockManager"
      hosts: "localhost:2181"

    orchestration:
      class: "pydiverse.pipedag.engine.SequentialEngine"
      ## Activate this class to work either with prefect 1.3.0 or prefect 2.0
      # class: "pydiverse.pipedag.engine.PrefectEngine"

instances:
  __any__:
    technical_setup: default
    # The following Attributes are handed over to the flow implementation (pipedag does not care)
    attrs:
      # by default we load source data and not a sampled version of a loaded database
      copy_filtered_input: false

  full:
    # Full dataset is using default database connection and schemas
    instance_id: pipedag_full
    # Run this instance under @pytest.mark.slow5
    tags: pytest_mark_slow5

  midi:
    # Full dataset is using default database connection and schemas
    instance_id: pipedag_midi
    attrs:
      # copy filtered input from full instance
      copy_filtered_input: true
      copy_source: full
      copy_per_user: false
      sample_cnt: 2  # this is just dummy input where we sample 2 rows

    # Run this instance under @pytest.mark.slow4
    tags: pytest_mark_slow4
    # Run only stage_2 under @pytest.mark.slow3
    stage_tags:
      pytest_mark_slow3:
        - simple_flow_stage2

  mini:
    # Full dataset is using default database connection and schemas
    instance_id: pipedag_mini
    attrs:
      copy_filtered_input: true
      copy_source: full
      copy_per_user: false
      sample_cnt: 1  # this is just dummy input where we sample 1 row

    # Run this instance under @pytest.mark.slow2
    tags: pytest_mark_slow2
    # Run only stage_2 under @pytest.mark.slow1
    stage_tags:
      pytest_mark_slow1:
        - simple_flow_stage2

  mssql:
    # Full dataset is using default database connection and schemas
    table_store:
      database_connection: mssql

flows:
#  __any__:
#    instances:
#      # it would be equivalent to move everything in "instances:" to here
  test_instance_selection:
    instances:
      full:
        schema_suffix: "_full"
    schema_prefix: "instance_selection_"
```

**Comment by windiana42** [comment]
The parsing of this pipedag configuration should ensure that the following anchor syntax will yield the exact same resulting pipedag-config-dictionary:
```yaml
name: pipedag_tests
strict_instance_lookup: true  # default value: true
_database_connections:
  postgres: &db_postgres
    # Postgres: this can be used after running `docker-compose up`
    url: "postgresql://sa:Pydiverse23@127.0.0.1:6543/{instance_id}"

  mssql: &db_mssql
    # SQL Server: this can be used after running `docker-compose up`
    url: "mssql+pyodbc://sa:PidyQuant27@127.0.0.1:1433/master?driver=ODBC+Driver+18+for+SQL+Server&encrypt=no"
    # schema_prefix: "master."  # SQL Server needs database.schema (uncomment only on of prefix and suffix)
    schema_prefix: "{instance_id}_"  # SQL Server needs database.schema (uncomment only on of prefix and suffix)
    schema_suffix: ".dbo"   # Alternatively SQL Server databases can be used as schemas with .dbo default schema

_blob_connections:
  file: &blob_file
    base_path: "/tmp/pipedag/blobs"

_technical_setups:
  default: &technical_setup_default
    # listen-interface for pipedag context server which synchronizes some task state during DAG execution
    network_interface: "127.0.0.1"
    # classes to be materialized to table store even without pipedag Table wrapper (we have loose coupling between
    # pipedag and pydiverse.transform, so consider adding 'pydiverse.transform.Table' in your config)
    auto_table: ["pandas.DataFrame", ]
    # abort as fast a possible on task failure and print most readable stack trace
    fail_fast: true

    instance_id: pipedag_default
    table_store:
      # Postgres: this can be used after running `docker-compose up`
      <<: *db_postgres
      create_database_if_not_exists: True

      ## SQL Server: this can be used after running `docker-compose up`
      #<<: *db_mssql

      class: "pydiverse.pipedag.backend.table.SQLTableStore"

      # print select statements before being encapsualted in materialize expressions and tables before writing to
      # database
      print_materialize: true
      # print final sql statements
      print_sql: true

    blob_store:
      class: "pydiverse.pipedag.backend.blob.FileBlobStore"
      <<: *blob_file

    lock_manager:
      class: "pydiverse.pipedag.backend.lock.ZooKeeperLockManager"
      hosts: "localhost:2181"

    orchestration:
      class: "pydiverse.pipedag.engine.SequentialEngine"
      ## Activate this class to work either with prefect 1.3.0 or prefect 2.0
      # class: "pydiverse.pipedag.engine.PrefectEngine"

_instances: &instances
  __any__:
    <<: *technical_setup_default
    # The following Attributes are handed over to the flow implementation (pipedag does not care)
    attrs:
      # by default we load source data and not a sampled version of a loaded database
      copy_filtered_input: false

  full:
    # Full dataset is using default database connection and schemas
    instance_id: pipedag_full
    # Run this instance under @pytest.mark.slow5
    tags: pytest_mark_slow5

  midi:
    # Full dataset is using default database connection and schemas
    instance_id: pipedag_midi
    attrs:
      # copy filtered input from full instance
      copy_filtered_input: true
      copy_source: full
      copy_per_user: false
      sample_cnt: 2  # this is just dummy input where we sample 2 rows

    # Run this instance under @pytest.mark.slow4
    tags: pytest_mark_slow4
    # Run only stage_2 under @pytest.mark.slow3
    stage_tags:
      pytest_mark_slow3:
        - simple_flow_stage2

  mini:
    # Full dataset is using default database connection and schemas
    instance_id: pipedag_mini
    attrs:
      copy_filtered_input: true
      copy_source: full
      copy_per_user: false
      sample_cnt: 1  # this is just dummy input where we sample 1 row

    # Run this instance under @pytest.mark.slow2
    tags: pytest_mark_slow2
    # Run only stage_2 under @pytest.mark.slow1
    stage_tags:
      pytest_mark_slow1:
        - simple_flow_stage2

  mssql:
    # Full dataset is using default database connection and schemas
    table_store:
      database_connection: mssql

flows:
  __any__:
    instances: *instances
  test_instance_selection:
    instances:
      full:
        schema_suffix: "_full"
    schema_prefix: "instance_selection_"
```

**Comment by windiana42** [comment]
I suggest the loader of this config to look like this:
```python
class PipedagConfig:
    _default_config = None

    def __init__(self, pipedag_config_dict: dict[str, Any], config_file: str):
        self.pipedag_config_dict = pipedag_config_dict
        # TODO: provide good error messages about configuration with line number
        self.config_file = config_file

    @classmethod
    def load(cls, path: Path | str | None = None) -> PipedagConfig:
        if isinstance(path, str):
            path = Path(path)
        if path is None:
            if cls._default_config is not None:
                return cls._default_config
            else:
                path = find_config()
        else:
            if path.is_dir():
                path = find_config(dirs_to_check=[path])
        return load_config(path)
```

**Comment by windiana42** [comment]
You can use this loader in the following ways:
```
flow = create_flow1()
flow.run()  # will internally run cfg=PipedagConfig.load().get() 

cfg=PipedagConfig.load().get() # will load instance=__any__, flow_name=cfg.get_pipedag_name()
flow = create_flow2(cfg.flow_name, cfg.attrs)
flow.run(cfg)

with PipedagConfig.load().get(): # will load instance=__any__, flow_name=cfg.get_pipedag_name()
  flow = create_flow3()  # can get ConfigContext.get().get_pipedag_name() or ConfigContext.get().attrs
  flow.run()  # will work with cfg=ConfigContext.get()

cfg=PipedagConfig.load().get(flow_name="foo") # will load instance=__any__
flow = create_flow4(cfg.flow_name, cfg.attrs)
flow.run(cfg)
```
I would generally prefer the 2nd solution, but it is just fine to use one of the others.

**Comment by windiana42** [comment]
I started writing all this configuration complexity up:
https://github.com/pydiverse/pydiverse.pipedag/blob/cfg_instances/docs/reference/config.rst

**Comment by windiana42** [comment]
@NMAC427 I hope this commit reflects our discussion outcome: https://github.com/pydiverse/pydiverse.pipedag/commit/36be4b20411206744ed875e2cf2b8a4a1000f755

We wanted to remove the open() / close() concept altogether in favor of ConfigContext initializing non-serializable state with @cached_property. It is dropped while pickling and recreated from dictionary values on first access. We also renamed some get-functions to create-functions where only one caller asks to build an object like BaseLockManager or OrchestrationEngine and takes care of disposing it himself. We also renamed remaining close() functions to dispose() which can be used by the user as ConfigContext.dispose() to close connections manually. However, he then has to use PipedagConfig.get() again for further flow.run() calls.
